### ISSUE TITLE
Features of dependencies merge between different targets

### ISSUE THREAD

**Issue by samoylovfp** [rust-lang/cargo#6121]
TLDR;
Cargo.toml:
```
...
[target.t1.dependencies]
dep1 = { version = "1", features = ["f1"] }

[target.t2.dependencies]
dep1 = { version = "1", features = ["f2"] }
```
both `cargo build --target t1` and `cargo build --target t2` build dep1 with both f1 and f2 simultaneously

When same crate appears in different [target.*.dependencies] sections in Cargo.toml the features settings get lost.
Steps to reproduce:
```
$ rustup default nightly
$ cargo new test_cargo_confusion
$ cd test_cargo_confusion
$ echo '[target.wasm32-unknown-unknown.dependencies]
image = {version = "0.19", default-features = false, features = ["jpeg"]}' >> Cargo.toml
```
run either `cargo tree -p image --target=wasm32-unknown-unknown`
or `cargo build --target=wasm32-unknown-unknown` and observe "jpeg-decoder" but no "rayon" package.

$ echo '[target.literally_anything.dependencies]
image = "0.19"' >> Cargo.toml

run either `cargo tree -p image --target=wasm32-unknown-unknown`
or `cargo build --target=wasm32-unknown-unknown` and observe "rayon" crate building and jpeg-decoder being built with rayon support.

Desired behaviour:
when building for wasm32-unknown-unknown respect feature settings for "image" crate

This bug makes it impossible to use crate `image` in environments configured to produce builds both for wasm and  x64_86 since jpeg-decoder with rayon does not work in wasm32-unknown-unknown.
```
$ cargo --version
cargo 1.31.0-nightly (de314a8b2 2018-09-21)
```

**Comment by samoylovfp** [comment]
I have written a proof-of-concept test
https://github.com/samoylovfp/cargo/commit/388bac6baeb14668fdb0653a53877494f24ae560
Will try to dive into the code if I find time, desire and courage

**Comment by Eh2406** [comment]
@alexcrichton what is the intended behavior hear?

**Comment by samoylovfp** [comment]
This is exactly the problem, will close as a duplicate
